bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -441,12 +459,18 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
 
 PreProcessor::~PreProcessor() {
   LOG_TRACE(logger(), "~PreProcessor start");
-  msgLoopDone_ = true;
-  msgLoopSignal_.notify_all();
-  cancelTimers();
-  threadPool_.stop();
-  if (msgLoopThread_.joinable()) {
-    msgLoopThread_.join();
+
+  if (!msgLoopDone_) {
+    msgLoopDone_ = true;
+    msgLoopSignal_.notify_all();
+    cancelTimers();
+  }
+
+  if (!threadPool_.isStopped()) {
+    threadPool_.stop();
+    if (msgLoopThread_.joinable()) {
+      msgLoopThread_.join();
+    }
```

Comment:
I don't like the code duplication between the 'stop' function and the destructor. I think all the logic should be in the 'stop' function. Call it from the destructor as well.

Reply:
I think you should assert here.
A dtor should free up resources, not stop if you have a dedicated mechanism for stopping.. so if we have a dedicated stop mechanism, not being in 'stopped' state here means we have a programmer bug that we don't want to overcome in the dtor.
If a replica is not stopped here, means someone removed code from somewhere else and you want to know about it immediately and crash, not continue successfully (probably without knowing, since the programmer doesn't look at logs, especially not at shutdown).

Reply:
will do

Reply:
done